src/util/JoinAlgorithms/JoinAlgorithms.h

```diff
@@ -924,32 +1072,68 @@ struct BlockZipperJoinImpl {
         getFirstBlock(currentBlocksRight, currentEl);
 
     compatibleRowAction_.setInput(fullBlockLeft.get(), fullBlockRight.get());
-    auto addRowIndex = [begL = fullBlockLeft.get().begin(),
-                        begR = fullBlockRight.get().begin(),
-                        this](auto itFromL, auto itFromR) {
+    auto begL = fullBlockLeft.get().begin();
+    auto begR = fullBlockRight.get().begin();
+    auto addRowIndex = [&begL, &begR, this](auto itFromL, auto itFromR) {
       compatibleRowAction_.addRow(itFromL - begL, itFromR - begR);
     };
 
     auto addNotFoundRowIndex = [&]() {
       if constexpr (DoOptionalJoin) {
-        return [begL = fullBlockLeft.get().begin(), this](auto itFromL) {
+        return [this, begL = fullBlockLeft.get().begin(),
+                &subrangeRight](auto itFromL) {
+          if constexpr (potentiallyHasUndef) {
+            AD_CORRECTNESS_CHECK(!isUndefined_(subrangeRight.front()) &&
```

Comment:
As this might be in a somewhat tight loop, precompute the condition, and check after the handling of the block or what not else, that the lambda wasn't called etc. or just make this check `AD_EXPENSIVE_CHECK`